components/server/src/workspace/workspace-deletion-service.ts

```diff
@@ -94,8 +99,81 @@ export class WorkspaceDeletionService {
      * @param ws
      * @param includeSnapshots
      */
-    protected async deleteWorkspaceStorage(ws: WorkspaceAndOwner, includeSnapshots: boolean): Promise<boolean> {
-        await this.storageClient.deleteWorkspaceBackups(ws.ownerId, ws.id, includeSnapshots);
+    protected async deleteWorkspaceStorage(
+        ctx: TraceContext,
+        ws: WorkspaceAndOwner,
+        includeSnapshots: boolean,
+    ): Promise<boolean> {
+        const span = TraceContext.startSpan("deleteWorkspaceStorage", ctx);
+        try {
+            await this.storageClient.deleteWorkspaceBackups(ws.ownerId, ws.id, includeSnapshots);
+            let vss = await this.db
+                .trace({ span })
+                .findVolumeSnapshotForGCByWorkspaceId(ws.id, this.config.workspaceGarbageCollection.chunkLimit);
+            await Promise.all(vss.map((vs) => this.garbageCollectVolumeSnapshot({ span }, vs)));
+        } catch (err) {
+            TraceContext.setError({ span }, err);
+            throw err;
+        } finally {
+            span.finish();
+        }
         return true;
     }
+
+    /**
+     * Perform deletion of volume snapshot from all clusters and from cloud provider:
+     *  - throws an error if something went wrong during deletion
+     *  - returns true in case of successful deletion
+     * @param ctx
+     * @param vs
+     */
+    public async garbageCollectVolumeSnapshot(ctx: TraceContext, vs: VolumeSnapshot): Promise<boolean> {
+        const span = TraceContext.startSpan("garbageCollectVolumeSnapshot", ctx);
+
+        try {
+            const allClusters = await this.workspaceManagerClientProvider.getAllWorkspaceClusters();
+            // we need to do two things here:
+            // 1. we want to delete volume snapshot object from all workspace clusters
+            // 2. we want to delete cloud provider source snapshot
+            let wasDeleted = false;
+            let index = 0;
+
+            let availableClusters = allClusters.filter((c) => c.state === "available");
+            for (let cluster of availableClusters) {
+                const client = await this.workspaceManagerClientProvider.get(cluster.name);
+                const req = new DeleteVolumeSnapshotRequest();
+                req.setId(vs.id);
+                req.setVolumeHandle(vs.volumeHandle);
+
+                let softDelete = true;
```

Comment:
nit: Already mentioned it, but just to highlight: Especially this part feels a tad brittle, and it would be nice to find a better way to handle/encapsulate that logic.